core/src/commonJvmMain/kotlin/com/ustadmobile/core/torrent/UstadTorrentManagerImpl.kt

```diff
@@ -0,0 +1,138 @@
+package com.ustadmobile.core.torrent
+
+import com.turn.ttorrent.client.*
+import com.turn.ttorrent.client.storage.FairPieceStorageFactory
+import com.ustadmobile.core.account.Endpoint
+import com.ustadmobile.core.io.ext.UstadFileCollectionStorage.createFileCollectionStorage
+import com.ustadmobile.core.util.DiTag
+import kotlinx.coroutines.*
+import org.kodein.di.DI
+import org.kodein.di.direct
+import org.kodein.di.on
+import org.kodein.di.instance
+import org.kodein.di.instanceOrNull
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.io.File
+
+
+interface CommunicationManagerListener {
+
+    fun onCommunicationManagerChanged(
+            meteredNetwork: Boolean
+    )
+}
+
+class UstadTorrentManagerImpl(
+        val endpoint: Endpoint,
+        override val di: DI
+) : UstadTorrentManager, CommunicationManagerListener {
+
+    private val torrentDir: File by di.on(endpoint).instance(tag = DiTag.TAG_TORRENT_DIR)
+
+    private val containerDir by di.on(endpoint).instance<File>(tag = DiTag.TAG_DEFAULT_CONTAINER_DIR)
+
+    private var containerInfoHashMap = mutableMapOf<Long, String>()
+
+    private var activeTorrentDownloadMap = mutableMapOf<Long, ActiveTorrentInfo>()
+
+ /*   private var containerDownloadPathMap = mutableMapOf<Long, String>()
+
+    private var containerManagerMap = mutableMapOf<Long, TorrentManager>()
+
+    private var containerListenerMap = mutableMapOf<Long, DownloadListenerAdapter>()*/
+
+    // on android, di uses connectionManager and a provider to get the current communicationManager
+    // on jvm, communicationManager is a singleton
+    private val communicationManager: UstadCommunicationManager?
+        get() = di.direct.instanceOrNull()
+
+    // lock add and removeTorrent
+    private val torrentLock = Mutex()
```

Comment:
Adding is locked only by things that call addTorrent (eg startSeeding). The lock should be called within addTorrent itself.